From b388423f4ba9e2a67b610945863866dd665a4385 Mon Sep 17 00:00:00 2001
From: "Lay, Kuan Loon" <kuan.loon.lay@intel.com>
Date: Thu, 28 Jul 2016 10:36:24 +0800
Subject: [PATCH 1/5] mqtt: fix connect to broker failed

mosquitto_connect_async() will do connect() and write("CONNECT") to broker.
Get the following result:
a) connect() return "Operation now in progress"
b) write("CONNECT") return "Resource temporarily unavailable"

When using asynchronous socket, we need to wait for
"POLLOUT (Writing now will not block) event" then only can do write().

Now sol_mqtt_event_loop() callback can be invoke when the "POLLOUT event" happen.
Then the mosquitto_loop_write() in sol_mqtt_event_loop() will continue the
write("CONNECT") that is failed in the first attempt.

Upstream-Status: Accepted

Signed-off-by: Lay, Kuan Loon <kuan.loon.lay@intel.com>
Signed-off-by: Chai, Chong Yi <chong.yi.chai@intel.com>
---
 src/lib/comms/sol-mqtt-impl-mosquitto.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/lib/comms/sol-mqtt-impl-mosquitto.c b/src/lib/comms/sol-mqtt-impl-mosquitto.c
index d6d1a38..fea5ef5 100644
--- a/src/lib/comms/sol-mqtt-impl-mosquitto.c
+++ b/src/lib/comms/sol-mqtt-impl-mosquitto.c
@@ -419,7 +419,7 @@ sol_mqtt_connect(const struct sol_mqtt_config *config)
         goto error;
     }
 
-    mqtt->socket_watch = sol_fd_add(mqtt->socket_fd, SOL_FD_FLAGS_IN | SOL_FD_FLAGS_PRI,
+    mqtt->socket_watch = sol_fd_add(mqtt->socket_fd, SOL_FD_FLAGS_IN | SOL_FD_FLAGS_PRI | SOL_FD_FLAGS_OUT,
         sol_mqtt_event_loop, mqtt);
     SOL_NULL_CHECK_GOTO(mqtt->socket_watch, error);
 
-- 
1.9.1

